### PR TITLE
376 program table

### LINKED ISSUE #376: Program table
Programs:
- creation date
- program patient id

- [x] create program table
- [x] update table when programs are upserted
- [x] use table for program queries

### PR DESCRIPTION
**Pull request by clemens-msupply**
Defines a program enrollment table holding a list of programs patients are enrolled in.

When mutating a program document this table is updates (same as for patients).

Update ProgramNodes and add an additional programs endpoints.

Just thinking about to give the `program` table/endpoint a better name. `program` might be a bit confusing since it is about program enrollments and not the main program. The program enrollment specific patient id `program_patient_id` could then renamed to `enrollment_id` as well. Any thoughts?

Closes #376

### CODE REVIEW

File: server/repository/src/db_diesel/program_row.rs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+#[table_name = "program"]
+pub struct ProgramRow {
+    /// The row id
+    pub id: String,
+    /// The type of the program, same as the matching program document type.
+    #[column_name = "type_"]
+    pub r#type: String,
+    /// The program document name
+    pub name: String,
+    /// The patient this program belong to
```
Comment (mark-prins) [inline server/repository/src/db_diesel/program_row.rs:29]: :nit - this is very nit-picky, sorry.
```suggestion
    /// The patient this program belongs to
```

**Review by mark-prins — APPROVED** [review]
👍 LGTM!

File: server/graphql/document/src/lib.rs [hunk 2]
```diff
@@ -104,6 +104,15 @@ impl DocumentQueries {
     ) -> Result<PatientSearchResponse> {
         patient_search(ctx, store_id, input)
     }
+
+    pub async fn programs(
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply server/graphql/document/src/lib.rs:108]: Its not really a join table but more an auxiliary table for program documents. But I think `enrollment` or `program_enrollment` would be a better name, will rename in another PR. There is currently no extra programs table planned, information about programs are stored in the `document_registry`, e.g. which document is a Program, Patient, Encounter and how they are related.

**Comment by clemens-msupply** [comment]
thanks for the review / finding my spelling mistakes!

**Pull request merged**